rs/sns/cli/src/upgrade_sns_controlled_canister.rs

```diff
@@ -255,6 +271,74 @@ pub struct UpgradeSnsControlledCanisterInfo {
     pub proposal_id: Option<ProposalId>,
 }
 
+pub async fn find_sns<C: CallCanisters>(
+    agent: &C,
+    caller_principal: PrincipalId,
+    target_canister_id: CanisterId,
+    target_controllers: BTreeSet<PrincipalId>,
+) -> Result<Option<Sns>, UpgradeSnsControlledCanisterError<C::Error>> {
+    let (user_controllers, canister_controllers): (Vec<_>, Vec<_>) =
+        target_controllers.into_iter().partition_map(|controller| {
+            if controller.is_self_authenticating() {
+                Either::Left(controller)
+            } else {
+                Either::Right(controller)
+            }
+        });
+
+    if user_controllers.contains(&caller_principal) {
```

Comment:
Seems like we also want to check that user_controllers is empty, no?  Do we actually care if the caller of this function also controls it?  Or just that there are self-authenticating controllers in charge of it?

Reply:
> Seems like we also want to check that user_controllers is empty, no?

We could do that, but checking that the SNS is sane is out of the minimal scope of this tool. So we could do it later, I think.

> Do we actually care if the caller of this function also controls it?

Not really; upgrading directly controlled canisters is out of scope of this tool. 

DFX already supports upgrading directly controlled canisters with large Wasms, in which case it makes more sense to use the target itself as the store. So we just hint at that DFX command and error out we detect this case.

Reply:
I don't understand, in that case, why we would care if _this user_ is the one controlling the canister to be upgraded.

SNSes actually take full control of canisters that they upgrade, and I think we do validation in the SNS. The canisters are added to the SNS via proposal, and then no other user can be in control